### ISSUE TITLE
User identifier from id_token

### ISSUE THREAD

**Issue by stevensajw** [Drenso/symfony-oidc#37]
Hi, 

I'm looking to connect to AzureAD OIDC and match users in an LDAP that don't know the sub. I am not able to use the upn claim as the `user_identifier_property`. The problem seems to be that the `upn` claim is not present in the userinfo endpoint of AzureAD, but only in the idtoken.  Apparently you can't change the data in de userinfo endpoint: https://learn.microsoft.com/en-us/azure/active-directory/develop/userinfo. Yay Microsoft :-(. 

_"Notes and caveats on the UserInfo endpoint
You can't add to or customize the information returned by the UserInfo endpoint.
To customize the information returned by the identity platform during authentication and authorization, use [claims mapping](https://learn.microsoft.com/en-us/azure/active-directory/develop/active-directory-claims-mapping) and [optional claims](https://learn.microsoft.com/en-us/azure/active-directory/develop/active-directory-optional-claims) to modify security token configuration."__

I don't see any existing option in the Drenso OIDC library to use data of the idtoken to use as `user_identifier_property`. I see some solutions: 

1. Add an explicit option that specifies the option to use the idtoken instead of userinfo for the  `user_identifier_property`.
2. Always merge the data from the idtoken into the OidcUserData jsonData 
`public function retrieveUserInfo(OidcTokens $tokens): OidcUserData`. If there is an id_token. The idtoken is already passed in inside $token, I think. 
3. Override `OidcClient` with a custom retrieveUserInfo()` which uses the id_token in the retrieveUserInfo. Which would work for me now and not require altering this library. 

What do you think is the way to do this best? Maybe I missed something obvious?

Regards, 
Albert-Jan

**Comment by stevensajw** [comment]
Maybe something like this would work? Get the data from the token and merge it with the existing data. Don't know if there are any edge cases (empty idtoken?) we would need to test. 

```php
 
  public function retrieveUserInfo(OidcTokens $tokens): OidcUserData
  {
    // Set the authorization header
    $headers = ["Authorization: Bearer {$tokens->getAccessToken()}"];

    // Retrieve the user information and convert the encoding to UTF-8 to harden for surfconext UTF-8 bug
    $jsonData = $this->urlFetcher->fetchUrl($this->getUserinfoEndpoint(), null, $headers);
    $jsonData = mb_convert_encoding($jsonData, 'UTF-8');

    // Read the data
    $userInfoData = json_decode($jsonData, true);

    // Get the data from the id_token
     $tokenData = $this->jwtHelper->decodeJwt($tokens->getIdToken(), 1);
     $tokenData = json_decode(json_encode($tokenData), true);
     $data = array_merge($tokenData, $userInfoData);
   
    // Check data due
    if (!is_array($data)) {
      throw new OidcException('Error retrieving the user info from the endpoint.');
    }

    return new OidcUserData($data);
  }
```

**Comment by bobvandevijver** [comment]
Well, what you're looking is indeed not possible with the current implementation, so you didn't miss something obvious. Only Microsoft did by leaving out the most important user information from the User Info endpoint 😄

To me option 1 seems best, as that would be the nicest point of integration in my opinion without processing the user info data. You can adjust the `OidcAuthenticator` to retrieve the `$userIdentifier` from either the `$userData` as it does now, or to retrieve it from the id token when configured. A new method to retrieve the identifier would be best.

I assume you still need access to the user info to be able to load the rest of the user information, right?

Are you up for a PR? You can look at how the `userIdentifierProperty`/`user_identifier_property` has been implemented, and adjust it for a new `useIdTokenForUserIdentifier`. The retrieval of the correct JWT part can be added to the `OidcJwtHelper` class, and maybe even refactor this line to use it: https://github.com/Drenso/symfony-oidc/blob/master/src/OidcClient.php#L445.

**Comment by stevensajw** [comment]
Hi Bob, 

That seems like fun. I'll try a PR, I have time this Wednesday, so will look into it then. 

Thanks, 

Albert-Jan

**Comment by bobvandevijver** [comment]
Fixed with #38, thank you @stevensajw 👍🏻

**Comment by stevensajw** [comment]
Learned a lot, thank you :-)
